File: sdks/typescript/src/utils/batch.ts
/**
 * Batch operation utilities
 *
 * Provides helpers for running batch operations with partial failure handling
 * using Promise.allSettled semantics.
 */

/**
 * Result of a single item in a batch operation
 */
export interface BatchItemResult<T> {
  /** Index of the item in the original array */
  index: number;
  /** Whether this item succeeded */
  success: boolean;
  /** The result if successful */
  data?: T;
  /** The error if failed */
  error?: string;
}

/**
 * Aggregated result of a batch operation
 */
export interface BatchResult<T> {
  /** Individual results for each item */
  results: BatchItemResult<T>[];
  /** Number of successful items */
  successCount: number;
  /** Number of failed items */
  failureCount: number;
  /** Total items processed */
  totalCount: number;
  /** Whether all items succeeded */
  allSucceeded: boolean;
}

/**
 * Process items in batch with partial failure handling.
 *
 * Uses Promise.allSettled internally so that one failure doesn't stop
 * other items from processing.
 *
 * @param items - Array of items to process
 * @param processor - Function to process each item
 * @returns Aggregated batch result with individual item results
 *
 * @example
 * ```typescript
 * const result = await batchProcess(
 *   [1, 2, 3],
 *   async (num) => {
 *     if (num === 2) throw new Error('Simulated failure');
 *     return num * 2;
 *   }
 * );
 *
 * console.log(result.successCount); // 2
 * console.log(result.failureCount); // 1
 * console.log(result.results[0].data); // 2
 * console.log(result.results[1].error); // 'Simulated failure'
 * ```
 */
export async function batchProcess<TInput, TOutput>(
  items: TInput[],
  processor: (item: TInput, index: number) => Promise<TOutput>
): Promise<BatchResult<TOutput>> {
  const settlements = await Promise.allSettled(
    items.map((item, index) => processor(item, index))
  );

  const results: BatchItemResult<TOutput>[] = settlements.map((settlement, index) => {
    if (settlement.status === 'fulfilled') {
      return {
        index,
        success: true,
        data: settlement.value,
      };
    } else {
      return {
        index,
        success: false,
        error: settlement.reason?.message || String(settlement.reason),
      };
    }
  });

  const successCount = results.filter((r) => r.success).length;
  const failureCount = results.length - successCount;

  return {
    results,
    successCount,
    failureCount,
    totalCount: results.length,
    allSucceeded: failureCount === 0,
  };
}

/**
 * Process items in batch with concurrency limit.
 *
 * Like batchProcess but limits how many items are processed concurrently.
 * Useful for rate-limited APIs or resource-constrained environments.
 *
 * @param items - Array of items to process
 * @param processor - Function to process each item
 * @param concurrency - Maximum concurrent operations (default: 10)
 * @returns Aggregated batch result with individual item results
 *
 * @example
 * ```typescript
 * const result = await batchProcessWithConcurrency(
 *   urls,
 *   async (url) => fetch(url).then(r => r.json()),
 *   5 // Only 5 concurrent requests at a time
 * );
 * ```
 */
export async function batchProcessWithConcurrency<TInput, TOutput>(
  items: TInput[],
  processor: (item: TInput, index: number) => Promise<TOutput>,
  concurrency: number = 10
): Promise<BatchResult<TOutput>> {
  const results: BatchItemResult<TOutput>[] = new Array(items.length);
  let currentIndex = 0;

  async function processNext(): Promise<void> {
    while (currentIndex < items.length) {
      const index = currentIndex++;
      const item = items[index];

      try {
        const data = await processor(item, index);
        results[index] = { index, success: true, data };
      } catch (error) {
        results[index] = {
          index,
          success: false,
          error: (error as Error).message || String(error),
        };
      }
    }
  }

  // Start workers up to concurrency limit
  const workers = Array(Math.min(concurrency, items.length))
    .fill(null)
    .map(() => processNext());

  await Promise.all(workers);

  const successCount = results.filter((r) => r.success).length;
  const failureCount = results.length - successCount;

  return {
    results,
    successCount,
    failureCount,
    totalCount: results.length,
    allSucceeded: failureCount === 0,
  };
}

File: sdks/python/xache/utils/batch.py
"""
Batch operation utilities

Provides helpers for running batch operations with partial failure handling.
"""

import asyncio
from dataclasses import dataclass, field
from typing import Any, Callable, Awaitable, Generic, List, Optional, TypeVar

T_Input = TypeVar("T_Input")
T_Output = TypeVar("T_Output")


@dataclass
class BatchItemResult:
    """Result of a single item in a batch operation."""
    index: int
    success: bool
    data: Any = None
    error: Optional[str] = None


@dataclass
class BatchResult:
    """Aggregated result of a batch operation."""
    results: List[BatchItemResult] = field(default_factory=list)
    success_count: int = 0
    failure_count: int = 0
    total_count: int = 0
    all_succeeded: bool = True


async def batch_process(
    items: List[Any],
    processor: Callable[..., Awaitable[Any]],
) -> BatchResult:
    """
    Process items in batch with partial failure handling.

    Uses asyncio.gather with return_exceptions so that one failure
    doesn't stop other items from processing.

    Args:
        items: Array of items to process
        processor: Async function(item, index) -> result

    Returns:
        Aggregated batch result with individual item results
    """
    tasks = [processor(item, i) for i, item in enumerate(items)]
    settlements = await asyncio.gather(*tasks, return_exceptions=True)

    results: List[BatchItemResult] = []
    for i, settlement in enumerate(settlements):
        if isinstance(settlement, Exception):
            results.append(BatchItemResult(
                index=i,
                success=False,
                error=str(settlement),
            ))
        else:
            results.append(BatchItemResult(
                index=i,
                success=True,
                data=settlement,
            ))

    success_count = sum(1 for r in results if r.success)
    failure_count = len(results) - success_count

    return BatchResult(
        results=results,
        success_count=success_count,
        failure_count=failure_count,
        total_count=len(results),
        all_succeeded=failure_count == 0,
    )


async def batch_process_with_concurrency(
    items: List[Any],
    processor: Callable[..., Awaitable[Any]],
    concurrency: int = 10,
) -> BatchResult:
    """
    Process items in batch with concurrency limit.

    Like batch_process but limits how many items are processed concurrently.
    Useful for rate-limited APIs or resource-constrained environments.

    Args:
        items: Array of items to process
        processor: Async function(item, index) -> result
        concurrency: Maximum concurrent operations (default: 10)

    Returns:
        Aggregated batch result with individual item results
    """
    semaphore = asyncio.Semaphore(concurrency)
    results: List[BatchItemResult] = [None] * len(items)  # type: ignore[list-item]

    async def process_item(index: int, item: Any) -> None:
        async with semaphore:
            try:
                data = await processor(item, index)
                results[index] = BatchItemResult(index=index, success=True, data=data)
            except Exception as e:
                results[index] = BatchItemResult(index=index, success=False, error=str(e))

    await asyncio.gather(*[process_item(i, item) for i, item in enumerate(items)])

    success_count = sum(1 for r in results if r.success)
    failure_count = len(results) - success_count

    return BatchResult(
        results=results,
        success_count=success_count,
        failure_count=failure_count,
        total_count=len(results),
        all_succeeded=failure_count == 0,
    )
